artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -1,35 +1,151 @@
+        Ok(offset)
+    } else if let Ok(offset_str) = unsafe { implicitly_convert_to_string(interp, &mut value) } {
+        let offset = Offset::try_from(offset_str)?;
+        Ok(offset)
+    } else {
+        Err(ArgumentError::with_message(
+            "+HH:MM, -HH:MM, UTC, A..I,K..Z, or a signed number of seconds expected for utc_offset",
+        )
+        .into())
+    }
+}
 
+// Check a Ruby Value Hash for the appropriate values to create and Offset
+// for a _time_.
+fn offset_from_options(interp: &mut Artichoke, options: Value) -> Result<Offset, Error> {
+    let hash: Vec<(Value, Value)> = interp.try_convert_mut(options)?;
+    let tz = hash
+        .into_iter()
+        .map(|(k, v)| {
+            let key = unsafe { Symbol::unbox_from_value(&mut k.clone(), interp)? }.bytes(interp);
+            if key == b"in" {
+                Ok(v)
+            } else {
+                Err(ArgumentError::with_message("unknown keyword"))?
+            }
+        })
+        .collect::<Result<Vec<Value>, Error>>()?;
+
+    match tz[..] {
+        [tz] => offset_from_value(interp, tz),
+        _ => Err(ArgumentError::with_message("unknown keyword"))?,
+    }
+}
+
+// Constructor
 pub fn now(interp: &mut Artichoke) -> Result<Value, Error> {
-    let now = Time::now();
+    let now = Time::now()?;
     let result = Time::alloc_value(now, interp)?;
     Ok(result)
 }
 
-pub fn at(interp: &mut Artichoke, seconds: Value, microseconds: Option<Value>) -> Result<Value, Error> {
+pub fn at(
+    interp: &mut Artichoke,
+    seconds: Value,
+    first: Option<Value>,
+    second: Option<Value>,
+    third: Option<Value>,
+) -> Result<Value, Error> {
+    // Coerce the params to the correct place. Specifically:
+    // - the options hash might not always be provided as the last argument.
+    // - subseconds can be provided with an optional symbol for the type of subsec.
+    //
+    // ```console
+    // [3.1.2] > Time.at(1)
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1)
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, in: "A")
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1, in: "A")
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec, in: "A")
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // ```
+    let (subsec, subsec_type, options) = match (
+        first.map(|opt| opt.ruby_type()),
+        second.map(|opt| opt.ruby_type()),
+        third.map(|opt| opt.ruby_type()),
+    ) {
+        (None, None, None) => (None, None, None),
+        (Some(Ruby::Hash), None, None) => (None, None, first),
+        (Some(Ruby::Fixnum), None, None) => (first, None, None),
+        (Some(_), None, None) => Err(ArgumentError::with_message("expected a number"))?,
```

Comment:
I think in this case we need to do the "implicitly convert to int" dance:

```
[3.1.2] > Time.at ""
<internal:timev>:272:in `at': can't convert String into an exact number (TypeError)
        from (irb):1:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > class X; def to_int; 5; end; end
=> :to_int
[3.1.2] > Time.at(X.new)
=> 1969-12-31 16:00:05 -0800
[3.1.2] > class Y; def to_int; ""; end; end
=> :to_int
[3.1.2] > Time.at(Y.new)
<internal:timev>:272:in `at': can't convert Y into an exact number (TypeError)
        from (irb):5:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > class Z; def to_int; 5.0; end; end
=> :to_int
[3.1.2] > Time.at(Z.new)
<internal:timev>:272:in `at': can't convert Z into an exact number (TypeError)
        from (irb):7:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
it looks like the implicit convert to int step is done below, should we collapse this match arm with the one above?

Reply:
I've done some fairly big changes to this part. Going to resolve this because of those changes